tests/models/test_default_evaluator.py

```diff
@@ -636,25 +646,110 @@ def example_custom_metric_with_artifacts(given_df, given_metrics):
-            "true_count": given_metrics["true_negatives"] + given_metrics["true_positives"],
-            "positive_count": np.sum(eval_df["prediction"]),
-        }
+def test_infer_artifact_type_and_ext_raise_exception_for_non_file_artifact_path(tmp_path):
+    with pytest.raises(
+        MlflowException,
+        match=f"produced an unsupported artifact 'non_file_artifact' with path '{tmp_path}'",
+    ):
+        _infer_artifact_type_and_ext("non_file_artifact", tmp_path, "")
+
 
+def test_infer_artifact_type_and_ext_raise_exception_for_unsupported_file_extension(tmp_path):
+    path = tmp_path / "invalid_ext_example.some_ext"
+    with open(path, "w") as f:
+        f.write("some stuff that shouldn't be read")
+    with pytest.raises(
+        MlflowException,
+        match=f"produced an unsupported artifact 'invalid_ext_artifact' with path '{path}'",
+    ):
+        _infer_artifact_type_and_ext("invalid_ext_artifact", path, "")
+
+
+@pytest.mark.parametrize(
+    "is_file,artifact,artifact_type,ext",
+    [
+        (True, lambda path: plt.figure().savefig(path), ImageEvaluationArtifact, "png"),
+        (True, lambda path: plt.figure().savefig(path), ImageEvaluationArtifact, "jpg"),
+        (True, lambda path: plt.figure().savefig(path), ImageEvaluationArtifact, "jpeg"),
+        (True, lambda path: json.dump([1, 2, 3], open(path, "w")), JsonEvaluationArtifact, "json"),
+        (
+            True,
+            lambda path: np.save(path, np.array([1, 2, 3]), allow_pickle=False),
+            NumpyEvaluationArtifact,
+            "npy",
+        ),
+        (
+            True,
+            lambda path: pd.DataFrame({"test": [1, 2, 3]}).to_csv(path, index=False),
+            CsvEvaluationArtifact,
+            "csv",
+        ),
+        (
+            True,
+            lambda path: pd.DataFrame({"test": [1, 2, 3]}).to_parquet(path),
+            ParquetEvaluationArtifact,
+            "parquet",
+        ),
+        (False, pd.DataFrame({"test": [1, 2, 3]}), CsvEvaluationArtifact, "csv"),
+        (False, np.array([1, 2, 3]), NumpyEvaluationArtifact, "npy"),
+        (False, plt.figure(), ImageEvaluationArtifact, "png"),
+        (False, {"a": 1, "b": "e", "c": 1.2, "d": [1, 2]}, JsonEvaluationArtifact, "json"),
+        (False, [1, 2, 3, "test"], JsonEvaluationArtifact, "json"),
+    ],
+)
+def test_infer_artifact_type_and_ext(is_file, artifact, artifact_type, ext, tmp_path):
+    if is_file:
+        artifact_representation = tmp_path / f"test.{ext}"
+        artifact(artifact_representation)
+    else:
+        artifact_representation = artifact
+    inferred_from_path, inferred_type, inferred_ext = _infer_artifact_type_and_ext(
+        f"{ext}_{artifact_type.__name__}_artifact", artifact_representation, ""
+    )
+    assert not (is_file ^ inferred_from_path)
+    assert inferred_type is artifact_type
+    assert inferred_ext == f".{ext}"
+
+
+def _start_run_for_custom_metric_tests(model_uri, dataset, custom_metrics):
```

Comment:
Nit: I'd rename this function to something like `_get_results_for_custom_metrics_tests` since the start run isn't the main point